modules/auxiliary/gather/manageengine_datasecurity_plus_xnode_enum.rb

```diff
@@ -0,0 +1,236 @@
+      aggr_max_query = { 'aggr' => { 'max' => { 'field' => 'UNIQUE_ID' } } }
+      res_code, res = get_response(@sock, action_dr_search(repo, ['UNIQUE_ID'], aggr_max_query))
+      aggr_max = process_dr_search(res, res_code, repo, ['UNIQUE_ID'], 'aggr_max')
+      # check if aggr_max is nil, as that means process_dr_search failed and we should skip to the next repo
+      next if aggr_max.nil?
+
+      print_good("Data repository #{repo} contains #{total_hits} records with ID numbers between #{aggr_min} and #{aggr_max}.")
+
+      repo_record_info_hash [repo] = {
+        'total_hits' => total_hits.to_i,
+        'aggr_min' => aggr_min.to_i,
+        'aggr_max' => aggr_max.to_i
+      }
+    end
+
+    # check if we found any repositories that contained any data
+    return if repo_record_info_hash.empty?
+
+    if dump_all
+      data_to_dump = ad_audit_plus_data_repos
+    else
+      data_to_dump = grab_config(config_file)
+
+      case data_to_dump
+      when config_status::CONFIG_FILE_DOES_NOT_EXIST
+        fail_with(Failure::BadConfig, "Unable to obtain the Xnode data repositories to target from #{config_file} because this file does not exist. Please correct your 'CONFIG_FILE' setting or set 'DUMP_ALL' to true.")
+      when config_status::CANNOT_READ_CONFIG_FILE
+        fail_with(Failure::BadConfig, "Unable to read #{config_file}. Check if your 'CONFIG_FILE' setting is correct and make sure the file is readable and properly formatted.")
+      when config_status::DATA_TO_DUMP_EMPTY
+        fail_with(Failure::BadConfig, "The #{config_file} does not seem to contain any data repositories and fields to dump. Please fix your configuration or set 'DUMP_ALL' to true.")
+      when config_status::DATA_TO_DUMP_WRONG_FORMAT
+        fail_with(Failure::BadConfig, "Unable to obtain the Xnode data repositories to target from #{config_file}. Check if your 'CONFIG_DIR' setting is correct or set 'DUMP_ALL' to true.")
+      end
+    end
+
+    # try and dump the database tables Xnode has access to
+    data_to_dump.each do |repo, fields|
+      if fields.blank? && !dump_all
+        print_error("Unable to obtain any fields for the data repository #{repo} to query. Skipping this table. Check your config file for this module if this is unintended behavior.")
+        next
+      end
+
+      # check if we actually found any records for the repo
+      next unless repo_record_info_hash.include?(repo)
+
+      total_hits = repo_record_info_hash[repo]['total_hits']
+      id_range_lower = repo_record_info_hash[repo]['aggr_min']
+      id_range_upper = id_range_lower + 9
+      max_id = repo_record_info_hash[repo]['aggr_max']
+      query_ct = 0
+
+      results = []
+      print_status("Attempting to request #{total_hits} records for data repository #{repo} between IDs #{id_range_lower} and #{max_id}. This could take a while...")
+      hit_upper_limit = false
+      loop do
```

Comment:
This code is present in both [manageengine_adaudit_plus_xnode_enum.rb](https://github.com/rapid7/metasploit-framework/pull/16725/files#diff-ab6eac7beea4596415f3e17fb07357ed960d018b08a5ec2e6a54293fca422810) and [manageengine_datasecurity_plus_xnode_enum.rb](https://github.com/rapid7/metasploit-framework/pull/16725/files#diff-b551acafa870c3d0ef819df55705023abc38a08329182d5891c9dfd44849ff64).

Boolean `hit_upper_limit` is defined immediately before the `loop` which also uses `hit_upper_limit` for flow control.

I suspect this loop could be cleaner, utilising `until hit_upper_limit do`.

Loop control with `hit_upper_limit` occurs towards the end of the loop:

```ruby
        if id_range_upper > max_id
          if hit_upper_limit
            results += partial_results unless partial_results.nil?
            break
          end
          hit_upper_limit = true
          id_range_upper = max_id
        end
        next if partial_results.nil?

        results += partial_results
      end
```

This code contains unnecessary duplication.

By using `until hit_upper_limit do`, this part of the loop could be simplified significantly:

```ruby
        if id_range_upper > max_id
          hit_upper_limit = true
          id_range_upper = max_id
        end

        results += partial_results unless partial_results.nil?
      end
```

This approach also has the advantage of removing a `break` from within nested loops (`data_to_dump.each` and `loop do`) which can be confusing.

I'm reading and reviewing your code in the GitHub UI. It is entirely possible that I made a mistake. Although if I've made a mistake that's a strong indication that the code is hard to read and should be refactored.

Reply:
Thanks! I ended up using `until hit_upper_limit` because rubocop didn't like `until hit_upper_limit do`. I didn't know you could use loop control like this, works like a charm. Had to make a few more tweaks to make sure it didn't omit the last range to query. I agree this was messy, so I'm happy with the improvements